storm-core/src/jvm/org/apache/storm/ui/resources/StormApiResource.java

```diff
@@ -0,0 +1,741 @@
+            e.printStackTrace();
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+
+        } catch (NotAliveException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        } catch (TException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * /api/v1/topology/summary -> topo history.
+     */
+    @GET
+    @Path("/topology/summary")
+    @Produces("application/json")
+    public Response getTopologySummary(@Context SecurityContext securityContext) {
+        try {
+            NimbusClient nimbusClient = NimbusClient.getConfiguredClient(Utils.readStormConfig());
+            return Response.status(Response.Status.OK).entity(JSONValue.toJSONString(
+                    UIHelpers.getAllTopologiesSummary(
+                            nimbusClient.getClient().getClusterInfo().get_topologies(),
+                            config
+                    ))).build();
+        } catch (AuthorizationException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+        } catch (TException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * /api/v1/topology -> topo.
+     */
+    @GET
+    @Path("/topology/{id}")
+    @Produces("application/json")
+    public Response getTopology(@PathParam("id") String id,
+                              @DefaultValue(":all-time") @QueryParam("window") String window,
+                              @QueryParam("sys") boolean sys) {
+        try {
+            NimbusClient nimbusClient = NimbusClient.getConfiguredClient(Utils.readStormConfig());
+            return Response.status(Response.Status.OK).entity(JSONValue.toJSONString(
+                    UIHelpers.getTopologySummary(
+                            nimbusClient.getClient().getTopologyPageInfo(
+                                    id, window, sys),
+                            window, config,
+                            servletRequest.getRemoteUser()
+                    ))).build();
+        } catch (AuthorizationException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.UNAUTHORIZED).build();
+
+        } catch (NotAliveException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        } catch (TException e) {
+            e.printStackTrace();
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * /api/v1/topology-workers/:id -> topo workers.
+     */
+    @GET
+    @Path("/topology-workers/{id}")
+    @Produces("application/json")
+    public Response getTopologyWorkers(@PathParam("id") String id) {
```

Comment:
We might want to test this API a few times.   In the clojure we were URL decoding the id again.  Not sure why, but this is something we should look into.